custom_function_double_backward.py

```diff
@@ -0,0 +1,206 @@
+    def backward(ctx, grad_out):
+        # A function support double backward automatically if autograd
+        # is able to record the computations performed in backward
+        x, = ctx.saved_tensors
+        return grad_out * 2 * x
+
+# Use double precision because finite differencing method magnifies errors
+x = torch.rand(3, 3, requires_grad=True, dtype=torch.double)
+torch.autograd.gradcheck(Square.apply, x)
+# Use gradcheck to verify second-order derivatives
+torch.autograd.gradgradcheck(Square.apply, x)
+
+######################################################################
+# We can use torchviz to visualize the graph to see why this works
+import torchviz
+
+x = torch.tensor(1., requires_grad=True).clone()
+out = Square.apply(x)
+grad_x, = torch.autograd.grad(out, x, create_graph=True)
+
+# We can see that the gradient wrt to x, is itself a function of x (dout/dx = 2x)
+# And the graph of this function has been properly constructed
+torchviz.make_dot((grad_x, x, out), {"grad_x": grad_x, "x": x, "out": out})
+
+
+######################################################################
+# A slight variation on the previous example is to save an output
+# instead of input. The mechanics are similar because outputs are also
+# associated with a grad_fn.
+class Exp(torch.autograd.Function):
+    # Simple case where everything goes well
+    @staticmethod
+    def forward(ctx, x):
+        # This time we save the output
+        result = torch.exp(x)
+        ctx.save_for_backward(result)
+        return result
+    
+    @staticmethod
+    def backward(ctx, grad_out):
+        result, = ctx.saved_tensors
+        return result * grad_out
+
+x = torch.tensor(1., requires_grad=True, dtype=torch.double).clone()
+# Validate our gradients using gradcheck
+torch.autograd.gradcheck(Exp.apply, x)
+torch.autograd.gradgradcheck(Exp.apply, x)
+
+out = Exp.apply(x)
+grad_x, = torch.autograd.grad(out, x, create_graph=True)
+torchviz.make_dot((grad_x, x, out), {"grad_x": grad_x, "x": x, "out": out})
+
+######################################################################
+# A more tricky case is when we need to save an intermediate result.
+# We demonstrate this case by implementing sinh(x) = (e^x - e^-x) / 2
+# Since the derivative of sinh is cosh, it might be useful to reuse
+# exp(x) and exp(-x), the two intermediate results in forward
+# in the backward computation.
+#
+# Intermediate results should not be directly saved and used in backward though.
+# Because forward is performed in no-grad mode, if an intermediate result
+# of the forward pass is used to compute gradients in the backward pass
+# the backward graph of the gradients would not include the operations
+# that computed the intermediate result. This leads to incorrect gradients.
+class Sinh2(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        expx = torch.exp(x)
+        expnegx = torch.exp(-x)
+        ctx.save_for_backward(expx, expnegx)
```

Comment:
Do we want the "easy trap" case where someone saves expx on the ctx and show the two graphs? With a big warning on the Function that is wrong of course :D

Reply:
The graph should be the same right? So maybe we don't need an entire separate implementation shown. I think its good to mention this pitfall in a comment though.

Reply:
No it won't 
In one, the Node will have only one output and only that will be used in backward.
In the second, you will see the two outputs and that both are actually properly linked in the double backward graph

Reply:
Tried it out here: https://colab.research.google.com/drive/1sSVpP1Dmn4tYIXTCUmWJP3ON9gC4cT_v
The only case where I see it breaking is `ctx.result = torch.exp(x); return torch.exp(x)` which seems a little odd to do haha

Reply:
Because your intermediary result is not an intermediary. I updated the notebook to make it fail.
And another custom function that is correct

Reply:
ohhh crap I was thinking about the Exp function, didn't realize the comment was actually about Sinh